src/ocean_emulators/models/modules/encoder.py

```diff
@@ -0,0 +1,90 @@
+# Sources inspired by the following implementations:
+# - https://github.com/microsoft/aurora/blob/main/aurora/model/patchembed.py
+# - https://github.com/lucidrains/vit-pytorch
+from typing import Any
+
+import torch
+from einops import rearrange
+from jaxtyping import Float
+from perceiver_pytorch import Perceiver
+from torch import nn
+
+from ocean_emulators.constants import Input
+
+
+class PerceiverEncoder(nn.Module):
+    """A perceiver-based encoder for Samudra's flattened data (a whole column of the ocean, with history).
+
+    Args:
+        n_channels (int): the number of input channels (roughly:  time x variable x (surface + depths)).
+        patch_size (int): the size of the patches to embed. Patches must evenly divide the input grid.
+        embed_dim (int): size of the latent dimension.
+        perceiver_depth (int): depth of the perceiver module core.
+        perceiver_kwargs (dict): keyword arguments to pass to the perceiver module. The docs for this module
+          can be found here: https://github.com/lucidrains/perceiver-pytorch#usage.
+    """
+
+    def __init__(
+        self,
+        n_channels: int,
+        patch_size: int | tuple[int, int],
+        embed_dim: int,
+        perceiver_depth: int,
+        **perceiver_kwargs: dict[str, Any],
+    ) -> None:
+        super().__init__()
+        self.n_channels = n_channels
+        if isinstance(patch_size, int):
+            self.patch_size: tuple[int, int] = (patch_size, patch_size)
+        else:
+            assert isinstance(patch_size, tuple) and len(patch_size) == 2, (
+                "Patch sizes must only span spatial dimensions (lat and lon)!"
+            )
+            self.patch_size = patch_size
+        self.embed_dim: int = embed_dim
+
+        self.norm_patches = nn.LayerNorm(self.n_channels)
+        self.perceiver = Perceiver(
+            num_freq_bands=4,
+            max_freq=1.0,
+            depth=perceiver_depth,
+            input_axis=2,  # Number of positional dims before token dim
+            input_channels=self.n_channels,  # input_dim
+            num_classes=embed_dim,  # output_dim
+            **perceiver_kwargs,
+        )
+        self.norm_embedding = nn.LayerNorm(embed_dim)
+
+    def forward(self, x: Input) -> Float[torch.Tensor, "*batch h w {self.embed_dim}"]:
```

Comment:
```suggestion
    def forward(self, x: Input) -> Float[torch.Tensor, "batch h w {self.embed_dim}"]:
```

The code below looks like it works for exactly 1 batch dimension, but `*batch` means "zero or more"